### ISSUE TITLE
DTLS 1.3

### ISSUE THREAD

**Issue by onovo1** [openssl/openssl#13900]
Hi!

Is there any plan to support DTLS 1.3?
Somebody asked the same question 4 years ago but at that time it wasn't any plan yet.

Thanks!

Oscar

**Comment by mattcaswell** [comment]
Still no plans. I understand it is still in draft. Not sure what the timeline is for RFC publication. Perhaps @kaduk knows.

**Comment by onovo1** [comment]
The specification has already been submitted to IESG for Publication.

This is the last version of the draft:

https://www.ietf.org/archive/id/draft-ietf-tls-dtls13-39.txt

**Comment by kaduk** [comment]
@onovo1 I am the IESG member to which it was submitted.
The draft's substate is "revised I-D needed", and https://github.com/tlswg/dtls13-spec/pulls shows many pending changes, including several that per https://mailarchive.ietf.org/arch/msg/tls/NppEqu8FnXUktgQX5Wy0_uF_NiA/ required some additional review from the WG.
The final RFC might take around three months from when a new I-D appears with those changes.

**Comment by onovo1** [comment]
Thank you @Kaduk for the information!

All pending changes has reviewed by the WG and merged into a new version of the document yesterday:
https://mailarchive.ietf.org/arch/msg/tls/aO-eaHBNTdRb8tYdEcLTUvjaSSU/

The new version of the draft can be found here:
https://www.ietf.org/archive/id/draft-ietf-tls-dtls13-40.txt

So, I understand this version could be the last before the RFC.
It would be nice to start planning for its implementation.

**Comment by Haidong988** [comment]
> The final RFC might take around three months from when a new I-D appears with those changes.

It seems that work has been done now as version-43, am I right?

**Comment by kaduk** [comment]
> > The final RFC might take around three months from when a new I-D appears with those changes.
> 
> It seems that work has been done now as version-43, am I right?

Yes.  Per https://datatracker.ietf.org/doc/draft-ietf-tls-dtls13/ the "IESG state" is 'RFC Ed Queue", and looking at the RFC Editor's queue at https://www.rfc-editor.org/current_queue.php DTLS 1.3 has been in the state "EDIT*R" for 2.9 weeks.  It's common to remain in an EDIT-related state for about 6-8 weeks.

**Comment by tambry** [comment]
DTLS 1.3 seems to have been published as [RFC 9147](https://datatracker.ietf.org/doc/rfc9147/).

**Comment by kaleb-himes** [comment]
Once this feature is available please feel free to use the wolfSSL DTLS 1.3 implementation for inter-op testing [wolfssl/src/dtls13.c](https://github.com/wolfSSL/wolfssl/blob/master/src/dtls13.c)!

Examples to make setup and testing easy can be found here:
[client-dtls13.c](https://github.com/wolfSSL/wolfssl-examples/blob/master/dtls/client-dtls13.c)
[server-dtls13.c](https://github.com/wolfSSL/wolfssl-examples/blob/master/dtls/server-dtls13.c)

[Documentation](https://www.wolfssl.com/whats-new-dtls-1-3/)

**Comment by fwh-dc** [comment]
I am interested in this feature. I am not sure about the complexity of adding the feature, but would you accept a contribution for this feature?

**Comment by mattcaswell** [comment]
In principle yes, but my expectation is that this will be difficult to do in practice and would be a *significant* amount of work.

On the on the one hand we already have DTLSv1.2 and TLSv1.3 support, so this is "just" extending what is already there to implement DTLSv1.3. However TLSv1.3 has impacts extensively throughout libssl, so everywhere that there is a check for `SSL_CONNECTION_IS_TLS13()` will have to be examined to see what impacts DTLS has. I also expect some significant changes in the record layer implementation (ssl/record/methods/tls13_meth.c).

I also would expect significant additional tests to cover the various DTLS cases.

Added to this is that we are expecting to move to time based releases (i.e. fixed time with movable scope). I would expect DTLSv1.3 to require several PRs to complete. An implication of time based releases is that master needs to be ready-to-release at any point (we cannot have "half finished" features in the master branch in case they don't get finished by the time we do the next release). So, we are expecting to start using "feature branches" for such large features - which we have not done before (and have definitely never done for an external contribution). I expect teething problems here while we figure out how this is supposed to work.

Any work for DTLSv1.3 should have a design written for it (see doc/designs/quic-design for examples), and I would recommend discussing it with the OTC before embarking on large scale work.

Review effort required is likely to be significant and it is unclear where that would fit in the priorities list and when it would be done.

I'm saying all of this not to put you off, but just to be clear about the level of effort this is likely to require. It is not unusual for large features to take >1 year to get into the codebase (e.g. see #17172 , #9155 - later replaced by #18186, #16620 - later replaced by #18185, etc). This particular feature is probably more complex than any of those.

**Comment by fwh-dc** [comment]
Thanks for the detailed answer.

It's good to know that it is not impossible to have a contribution accepted.
I still need to check if I have time to work on the feature. I'll let you know if I have something to contribute.

**Comment by fwh-dc** [comment]
### For the OTC:

As requested by @mattcaswell here is my initial plan for developing the feature.

I notice another DTLS 1.3 feature request here: https://github.com/openssl/openssl/issues/4804 . Maybe that should be closed as a duplicate to make sure all relevant information goes here.

**Functionality:**
I want to develop an “MVP” of DTLS. Hence it will support all “MUST”’s of the RFC but not necessarily the “SHOULD”'s.
I’ve compiled a list of discrepancies between DTLS 1.3 and TLS 1.3 that are not part of older dtls versions and which I intend to use as a development checklist:

* As with other versions of DTLS, DTLS 1.3 adds epoch and sequence numbers to records to handle different scenarios of record loss and reordering. And there are fields to handle record fragmentation.
* DTLSCiphertext omits the superfluous “legacy_version” and type fields of TLSCiphertext.
* DTLSCiphertext reduces the number of bits used for epoch and sequence numbers compared to the previous DTLS versions.
* DTLSCiphertext protects the sequence number with encryption. 
* DTLSCiphertext structure has a variable-length header.
* DTLSPlaintext sets the DTLS epoch to the least significant 2 octets of the connection epoch and the low order 48 bits of the 64 bit sequence number.
* ACK content type messages
* ChangeCipherSpec messages must not be sent I.e. middle box-compatibility is not supported.
* Must not echo legacy_session_id
* Must not use HelloVerifyRequest 
* EndOfEarlyData message is deprecated 

All features for enabling/negotiating Connection ID will not be supported.
The variable length header will be statically forced to use a 16-bit sequence id and a 16-bit length field and omit the connection id field.

Let me know if you see anything missing.

**Documentation:**
I will use the documentation of QUIC TLS as a reference, but my initial thought would be to document the data structures I add to openssl and to go through the DTLS 1.3 RFC and document how “SHOULD”’s and “MAY”’s are implemented. Each project typically have a different view of what is interesting to document. If you have a description of documentation requirements please let me know.

**Development process:**
For a start I will use the s_server and s_client to get the “sunshine” handshake functionality working by manually validating records. Later I need to figure out how I can test recovery of record loss, handshake retry requests etc. I assume you do something similar in your test code.
I will also test integration with the WolfSSL implementation.

_Q for OTC:_ Do you have some advice on testing that I need to know?
_Q for OTC:_ How do I get in contact with you if I need some advice on how to “bend” the existing implementation properly?

**Incremental PR’s:**
When developing the feature I will likely stumble upon code changes (refactoring/bugfixes) that are not directly related to the feature implementation and for such changes I’d like to submit PR’s. 

_Q for OTC:_ Is it ok to submit a PR for such code changes without raising an issue? I will ofc add a reasonable explanation to the PR.

**DTLS 1.3 PR:**
I will maintain a personal branch for DTLS1.3 development. I would like to maintain a single commit that I rebase on top of master once in a while to keep up to date. Unless the OTC oppose against that plan?

_Q for OTC:_ Would you like me to open a draft PR to allow you to follow the development of the feature?

**Comment by mattcaswell** [comment]
Responses from OTC:


> Q for OTC: Do you have some advice on testing that I need to know?

OTC: There are several different types of testing for DTLSv1.2 that we are currently using that you may need to update and/or extend as appropriate. We suggest that you review the various existing tests to get an understanding of what they cover. In particular:

- ssl-test-new. These tests configure a client and a server with numerous different config options and confirm that a connection succeeds or fails as expected with appropriate output. The "driver" for the tests is in test/ssl_test.c, and the various configurations are in the numerous test/ssl-tests/*.cnf files. The cnf files themselves are autogenerated from cnf.in files.

- dtlstest.c: Various miscellaneous DTLS tests. In particular tests for dropped and out-of-order records as well as DTLSv1_listen().

- dtlsv1listentest.c - More testing of DTLSv1_listen()

- dtls_mtu_test.c - Tests related to MTU, e.g. checking MTU limits are enforced

- bad_dtls_test.c - You probably don't need to touch this since it is related to testing of a pre-standardisation version of DTLS.

- TLSProxy tests - There are numerous TLSv1.3 related tests based on TLSProxy (a perl based testing framework). It allows us to introduce "invalid" protocol messages and confirm that we behave as expected. Unfortunately there is no DTLS support in this at the moment. (see test/recipes/70-test_sslrecords.t as an example)


Also you should consider interoperability testing (e.g. against WolfSSL)

> Q for OTC: How do I get in contact with you if I need some advice on how to “bend” the existing implementation properly?

OTC: Just add a comment to an existing issue or pull request asking for OTC attention. You can also ask questions in Github discussion items. Not all questions/comments necessarily need to be directed at the OTC as a whole. i.e. any OpenSSL team member should be able to answer, if knowledgeable about the particular topic. Only significant/major design issues need to be directed to OTC.


> Incremental PR’s:
> When developing the feature I will likely stumble upon code changes (refactoring/bugfixes) that are not directly related to the feature implementation and for such changes I’d like to submit PR’s.
>
> Q for OTC: Is it ok to submit a PR for such code changes without raising an issue? I will ofc add a reasonable explanation to the PR.

OTC: Yes, this is ok.

> DTLS 1.3 PR:
> I will maintain a personal branch for DTLS1.3 development. I would like to maintain a single commit that I rebase on top of master once in a while to keep up to date. Unless the OTC oppose against that plan?
>
> Q for OTC: Would you like me to open a draft PR to allow you to follow the development of the feature?

OTC: Please do *not* do everything in a single commit. In order for us to be able to review this we will need to see your changes in a series of easily digestible PRs. For this to work we will need to have a feature branch setup to receive your changes.


OTC: Some other comments:

- Please can you identify and provide details of any API changes that you think might be needed in order support this feature?

- Do you have have an approximate timeline over which you intend to do this work?

- We are not able to commit at this stage to including this work in any specific release

**Comment by fj-blanco** [comment]
> [fwh-dc](/fwh-dc)

Hi @fwh-dc,

I came across your plan for implementing DTLS 1.3 into OpenSSL, and I'm very interested in the project. I have been playing around testing some DTLS 1.3 features with WolfSSL, so perhaps I can help with this.

If you like the idea, feel free to reach out to me via email (which you can find in my GitHub bio) and I will be happy to help under your coordination.

Best.

**Comment by fwh-dc** [comment]
Hi @fj-blanco 

Thanks for the initiative! I'll message you straight away 👍

**Comment by fwh-dc** [comment]
@mattcaswell 
> OTC: There are several different types of testing… 

Thanks for a great overview!

> OTC: Just add a comment to an existing issue or pull request asking for OTC attention. You can also ask questions in Github discussion items. Not all questions/comments necessarily need to be directed at the OTC as a whole. i.e. any OpenSSL team member should be able to answer, if knowledgeable about the particular topic. Only significant/major design issues need to be directed to OTC.

Actually I think we can handle such “discussions” using your idea of a feature branch and incremental PR’s!

> OTC: Please do not do everything in a single commit. In order for us to be able to review this we will need to see your changes in a series of easily digestible PRs. For this to work we will need to have a feature branch setup to receive your changes.

Sure. Good point. Please do create the branch.

> * Please can you identify and provide details of any API changes that you think might be needed in order support this feature?

I’m not sure if I have the full overview. But at this stage I see no API changes (other than the ability to select DTLS 1.3). There might be some DTLS 1.3 configuration of the variable length header in the future, but it was my intention to make that static for now if that is okay with you?

> * Do you have have an approximate timeline over which you intend to do this work?

I cannot say if there will be other tasks that takes priority... But my original thought would be to develop this feature until the end of the year. Hopefully reaching a “useable” state.

> * We are not able to commit at this stage to including this work in any specific release

This is not a concern to us as this is developed as an experimental feature for evaluation purposes.

**Comment by mattcaswell** [comment]
> Sure. Good point. Please do create the branch.

We're still figuring out ourselves how feature branches are going to work. As I said above:

> So, we are expecting to start using "feature branches" for such large features - which we have not done before (and have definitely never done for an external contribution). I expect teething problems here while we figure out how this is supposed to work.

It looks like DTLSv1.3 might be the first thing that we use a feature branch for. For now, when you create your first PR for review, please just target the master branch. We won't actually merge it there - it will go to the new feature branch, but that branch will be based off of master branch anyway so this should not be a problem. It might take us a little while to actually create the new branch because we have to figure out how to manage it once created.

**Comment by pals-ucb** [comment]
@mattcaswell  Thanks for redirecting me here.
I am a student in UCB and few of us are working on using DTLS to evaluate PQC stack for IoT Devices.  We experimented with DTLS 1.2 over OQS/OpenSSL1.1, and made little progress.  We realized that OQS is moved on to OpenSSL3.0.
We also realized that DTLS 1.3 spec is almost their and decided put some effort in helping to build DTLS 1.3.

@fwh-dc I like your proposal and would like to join the group in helping to build the DTLS 1.3 on OpenSSL3. 
Please rope me into this effort and will be glad to contribute.

**Comment by fwh-dc** [comment]
@mattcaswell 

> We're still figuring out ourselves how feature branches are going to work.

Ah sorry for the misunderstanding! I will prepare some PR's. And await further instructions.

@pals-ucb 
Great that you show interest. I think it will be best to have the feature branch (that has been mentioned in this thread) up and running before we begin the collaboration. I'll be sure to let you know when we can progress.

**Comment by hannestschofenig** [comment]
Great to hear about this development. I would be interested to help.

**Comment by eabase** [comment]
What's the current status of this? 
I already commented in an old thread, so it may be a good idea to do some old issue maintenance and clean ups!

**Comment by paulidale** [comment]
No progress.
Not sure if it will be a priority for 3.3 or not.

A community submission would be considered.

**Comment by fwh-dc** [comment]
@eabase I will make some pull requests for adding dtls structs and relevant refactorings of the existing codebase this week. In order to collaborate we need a feature branch as described by Matt in an earlier comment. And as he explains there are some uncertainties on how that should handled.

**Comment by paulidale** [comment]
A community contribution would be view favourably IMO.

**Comment by fwh-dc** [comment]
Here’s an update on the development:

I’ve just finished breaking down some early development work. It is primarily focused on getting a TLS 1.3 handshake to run as DTLS. Many DTLS 1.3 specific features have not been developed yet.

Done so far:
* Updated relevant occurrences of SSL_CONNECTION_IS_TLS13() to support dtls 1.3. 
* Fixed some discrepancies between tls and dtls record functions. 
* Added some dtls 1.3 handling to tls13 specific functions.

I have created pull requests against master for all of the changes and have merged all the changes on to one branch: “dtls1.3-pr” (https://github.com/openssl/openssl/pull/22406) which serves as a starting point for further development. That branch will of course become obsolete when we have a feature branch up and running.

Please let me know if there's something i missed.

Here are some remaining tasks that others might be interested in developing:
* Support for writing the variable header to a record
* Support for reading the variable header from a record
* Support for the ACK message type
* Testing (Matt has posted an overview of different openssl test suites)

I know some have previously declared interest in contributing to the feature. Please confirm that you are still able to work on the feature and what you want to contribute such that we avoid duplicate work. 

Also do let me know if you have suggestions on how to proceed with the development.

When you have something to contribute and I will add it to the “dtls1.3-pr” branch.

**Comment by fj-blanco** [comment]
Great progress so far! I'll be diving into the 'Support for writing the variable header to a record' task this week. I'll give you any necessary feedback as I progress. If there's anything specific I should be aware of before getting started, let me know.

**Comment by fj-blanco** [comment]
Great progress so far! I'll be diving into the 'Support for writing the variable header to a record' task this week. I'll give you any necessary feedback as I progress. If there's anything specific I should be aware of before getting started, let me know.

**Comment by eabase** [comment]
@fwh-dc 
Did you get any help and further progress?

**Comment by fwh-dc** [comment]
> @fwh-dc Did you get any help and further progress?

No futher progress to share other than what's in this thread. I'm off on vacation for the next couple of weeks. After that I will have more time to work on this. I will be working on the design documentation of what has been done so far as the next item.

Please feel free to have a look at the current state and make a contribution.

**Comment by fj-blanco** [comment]
> @fwh-dc Did you get any help and further progress?

I have a first draft of the initial major changes required for the "Support for writing the variable header to a record" feature in this [fork](https://github.com/fj-blanco/openssl-dtls1.3). I want to dedicate more time to attempting tests on these changes before submitting a PR to fwh-dc:dtls1.3-pr. If anyone is interested in contributing or needs these changes to progress with other features, please reach out.

**Comment by nhorman** [comment]
The review work for DTLS1.3 on @fwh-dc patch series is being tracked here https://github.com/openssl/project/issues/594 with a subsequent task to merge it anticipated for 3.5.  Can this be closed as a duplicate of that issue?

**Comment by eabase** [comment]
Hi Neil, @nhorman 
I don't see why you want to close this issue with all the above comments, when you still have the project (you linked) open, that is independent of this? Please keep this open until fully implemented in public main.

**Comment by nhorman** [comment]
I think you answered your own question.  This issue asks a question about dtls1.3, and I've noted we have a tracking issue for 3.4 and another likely for 3.5 to complete the work. As the question is answered, I see no purpose in this issue remaining open.

I'm happy to change that though, but to do so I need a clear definition of action for this issue that is  independent of the tracking issue (that is to say, it needs to have an action associated with it).  Lacking that, it's not an issue, it's a discussion, which I'm also happy to convert it to.

**Comment by eabase** [comment]
Hi Neil, 
IDK what kind of weird project management process you have, but I think the only *action* here is to make openssl support `DTLS 1.3` as in the list of items you linked above. I would refrain from converting these old tickets to something else that effectively change (?) their original URL. Things and useful discussions often gets lost that way.

**Comment by fwh-dc** [comment]
Here’s a summary of the development activities remaining from the DTLS 1.3 feature implementation:

Work in progress:
* ACK record handling
* Sequence number encryption
* Correct DTLS 1.3 epochs
* DTLS 1.3 Downgrade mechanism
* DTLSv1_listen support for DTLS 1.3

Features that are up for grabs:
* Finish the implementation of the "variable header length” DTLS 1.3 specific feature (some has already been developed in https://github.com/fwh-dc/oqs-openssl/pull/3)
* Develop an interop test against wolfssl which supports DTLS 1.3.
* Handle interop issues from the test against wolfssl
* Additional DTLS 1.3 testing. We can probably reuse some existing test cases for tls 1.3 and older dtls versions.
* Fixing todos: TODO(DTLSv1.3)

The work is continuously added to https://github.com/openssl/openssl/tree/feature/dtls-1.3

I've been trying to keep notes on the current implementation in https://github.com/openssl/openssl/blob/feature/dtls-1.3/doc/designs/dtlsv1_3/dtlsv1_3-main.md

Contributions are very welcome.
Feel free to reach out if you are interested in contributing and need some more info to get started.

**Comment by Frosne** [comment]
We (Firefox) support DTLS1.3. 

If you're willing to, whenever you're ready, we will be happy to have an interoperability testing.

**Comment by eabase** [comment]
I see that you (open SSL) have leaft out Connection ID, which should be considereed essential for DTLS1.3 support!

> DTLSv1.3 connection id
> OpenSSL does not support Connection ID's (RFC 9146). Notably Openssl DTLSv1.3 clients will not offer the "connection_id" extension even though RFC 9147 states:  
*"DTLS clients which do not want to receive a Connection ID SHOULD still offer the "connection_id" extension [RFC9146]..."*

What's going on there?

**Comment by fwh-dc** [comment]
> I see that you (OpenSSL) have left out Connection ID, which should be considered essential for DTLS1.3 support!
> 
> > DTLSv1.3 connection id
> > OpenSSL does not support Connection ID's (RFC 9146). Notably Openssl DTLSv1.3 clients will not offer the "connection_id" extension even though RFC 9147 states:
> > _"DTLS clients which do not want to receive a Connection ID SHOULD still offer the "connection_id" extension [RFC9146]..."_
> 
> What's going on there?

I'm developing DTLS 1.3 as a contribution to OpenSSL to be able to integrate quantum algorithms in to an application that uses DTLS-SRTP.

Since DTLS 1.3 is a rather big contribution on its own I'd like to keep it as small as possible.

If someone wants to contribute Connection-ID support I see that as a feature on its own. Maybe the @openssl/otc can share some guidance here?

**Comment by mattcaswell** [comment]
Right. This initial effort is to get us a "minimum viable" DTLSv1.3 implementation. It does not preclude (and indeed I very much expect) future work to further extend it.

**Comment by seschu** [comment]
Hello there,

is my understanding correct that currently also the above mentioned feature branch will not be able to use the oqs provider via the groups feature? in DTLS1.3?

**Comment by t8m** [comment]
> Hello there,
> 
> is my understanding correct that currently also the above mentioned feature branch will not be able to use the oqs provider via the groups feature? in DTLS1.3?

I do not see why it would not be able to be used.

**Comment by mattcaswell** [comment]
TLS groups from an external provider only work for TLS protocol versions that they are declared to work for by that provider. It looks to me like the oqs-provider groups do not declare support for DTLSv1.3 (probably because there is no such protocol version in mainline OpenSSL). It should be a straightforward oqs-provider change to fix that, but I don't think it will be usable without that change.

https://github.com/open-quantum-safe/oqs-provider/blob/a4cfbc93e98cdf118a23ce1750f4a301b385b180/oqsprov/oqsprov_capabilities.c#L25-L110

@baentsch might want to comment further.

**Comment by baentsch** [comment]
This comment is correct: There's simply no DTLS1_3_VERSION #define `oqsprovider` can currently refer to.

For the fun of it, I just added it to the capabilities as per the link above and also to the test suite at https://github.com/open-quantum-safe/oqs-provider/blob/a4cfbc93e98cdf118a23ce1750f4a301b385b180/test/tlstest_helpers.c#L59-L62, and built using dtls-1.3 feature branch and all tests passed. Nearly too good to be true: @mattcaswell @t8m: Would you agree that just adding the "D" to the `SSL_CTX_set_min_proto_version` invocation parameters should "force" use of DTLS in these (handshake) tests?

@fwh-dc @seschu let me know if it'd help you if I'd push this onto a public "tracker-dtls-1.3" branch (to make this switchable and maybe add further DTLS 1.3-specific tests).

**Comment by seschu** [comment]
Hi @baentsch  and @t8m, thanks for your help first of all. I have to admit I do not quite get which changes to oqs provider are necessary so that I could test the functionality of the oqsprovider groups in dtls. 

tbh dtls1.2 would be fine it is just that my assumption is since groups were introduced with TLS 1.3 the same would be the case for dtls 1.3. What  am trying to do basically is just to conduct the tests for th eoqsprovider using s_client and s_server with the dilithium3 signature but using the -dtls1_2 (or -dtls1_3) option. Currently this fails for me. I am currently using oqsprovide rbased in this commit 7b21bd3bf35ba5fec53d5e5ab20f071b8afcd8b8

**Comment by seschu** [comment]
So just to be clear I am asking whether you could explain your answer again :) I am sorry

**Comment by fwh-dc** [comment]
> @fwh-dc @seschu let me know if it'd help you if I'd push this onto a public "tracker-dtls-1.3" branch (to make this switchable and maybe add further DTLS 1.3-specific tests).

I think that would be okay. But make it clear that it is by no means ready for production. I don't know which tests you do on the transport protocol level, but it might be "easy" to replicate some if not all existing test cases for TLS to DTLS.

**Comment by fwh-dc** [comment]
> So just to be clear I am asking whether you could explain your answer again :) I am sorry

I can try to explain:

First of all DTLS 1.3 is not fully implemented in OpenSSL, so if you are using a standard distribution you can only use up to DTLS 1.2.

The kem and signature algorithms in OQS-provider requires (D)TLS 1.3 in order to function.

Since DTLS 1.3 is not fully implemented for OpenSSL OQS-provider has no support for DTLS 1.3 (yet).

If @baentsch makes the tracker branch, that would allow you to do some experimental testing with a local build of OpenSSL ´feature/dtls-1.3´ branch.

**Comment by fwh-dc** [comment]
Yes I think would allow you to do experimental testing as well. Great that it works for you.

**Comment by seschu** [comment]
Just out of curiosity: What is missing in the dtls1.3 feature branch? is there any particular feature missing of dtls1.3?

**Comment by seschu** [comment]
@baentsch just to answer you as well: As far as I can tell this works so having this tracker branch might be a good idea

**Comment by mattcaswell** [comment]
> Would you agree that just adding the "D" to the SSL_CTX_set_min_proto_version invocation parameters should "force" use of DTLS in these (handshake) tests?

Not quite as simple as that unfortunately. You must also use the `DTLS_server_method()`/`DTLS_client_method` for creating the `SSL_CTX`. Additionally DTLS is datagram based so when setting BIOs you need to use a datagram based BIO.

**Comment by baentsch** [comment]
> Not quite as simple as that unfortunately. You must also use the DTLS_server_method()/DTLS_client_method for creating the SSL_CTX.

Thanks for that pointer, @mattcaswell . So now added -- and it immediately has an impact: DTLS1.3 tests begin failing for all Sphincs (aka SLH-DSA), CROSS and large-key Frodo and HQC variants but work for the other algorithms: See e.g., https://github.com/open-quantum-safe/oqs-provider/actions/runs/12238474126/job/34136746509 fyi @seschu @fwh-dc 

> Additionally DTLS is datagram based so when setting BIOs you need to use a datagram based BIO.

Is this really necessary? We're currently using mem_BIOs as we're in this testing not really interested in what shenenigans the network introduces, are we?

**Comment by eabase** [comment]
Is there an itemized tracking list somewhere for this?  
It would be great to have a clear overview where we are at...

Like a checked bullet list:

- [ ] Not done
- [x] This was done
- [ ] etc

**Comment by mattcaswell** [comment]
> Is this really necessary? We're currently using mem_BIOs as we're in this testing not really interested in what shenenigans the network introduces, are we?

This could cause unexpected breakage, or result in things passing that otherwise wouldn't. Recent versions of OpenSSL have `BIO_new_bio_dgram_pair` for this purpose (or alternatively `BIO_s_dgram_mem`). A normal mem BIO does not respect datagram boundaries, e.g. if you `BIO_read` half a datagram, then the correct semantics is that the remainder of the datagram should be dropped and the next `BIO_read` should move on to the next datagram - but a normal mem BIO does not do that.

**Comment by mattcaswell** [comment]
> This comment is correct: There's simply no DTLS1_3_VERSION #define oqsprovider can currently refer to.

Just to point out that oqsprovider could in theory define this value itself since it is a well know value defined by the RFC, e.g.

```` c
#if !defined(DTLS1_3_VERSION)
# define DTLS1_3_VERSION 0xFEFC
#endif
````
Doing so would mean oqsprovider could integrate DTLSv1.3 support in advance of it being available in mainline OpenSSL.

Whether that is desirable or not is another question...

**Comment by baentsch** [comment]
> Doing so would mean oqsprovider could integrate DTLSv1.3 support in advance of it being available in mainline OpenSSL.

Good proposal @mattcaswell , particularly considering the above mentioned actually required code changes in DTLS testing that seem sensible to land. Also this would avoid me maintaining a separate branch : I'm above all lazy :).

> Whether that is desirable or not is another question...

`oqsprovider` is meant to support "research and prototyping" (includes testing for me) and already contains a lot of code to that purpose -- primarily for PQC but why not for DTLS1.3?

So, I'm very open for that but leave this decision to @fwh-dc and/or @seschu : What's your preference/would it help you if we land DTLS1.3 (test) support in `oqsprovider` main?

**Comment by fwh-dc** [comment]
> So, I'm very open for that but leave this decision to @fwh-dc and/or @seschu : What's your preference/would it help you if we land DTLS1.3 (test) support in `oqsprovider` main?

It would be helpful to me since I am implementing DTLS 1.3 to be able to use it with oqsprovider. But I don't have time to do it  at the moment.

**Comment by fwh-dc** [comment]
> Is there an itemized tracking list somewhere for this? It would be great to have a clear overview where we are at...
> 
> Like a checked bullet list:
> 
> * [ ]  Not done
> * [x]  This was done
> * [ ]  etc

There's no list. I've been trying to keep an overview in https://github.com/openssl/openssl/blob/feature/dtls-1.3/doc/designs/dtlsv1_3/dtlsv1_3-main.md

If you are interested in contributing please let me know and I can give you an overview of outstanding issues. You can also have a look for `TODO(DTLSv1.3)` in the code.

**Comment by baentsch** [comment]
> It would be helpful to me since I am implementing DTLS 1.3 to be able to use it with oqsprovider. But I don't have time to do it at the moment.

NP. Will try to do it: Probably unconditionally adding the baseline capability and an option to activate DTLS/datagram testing as per the suggestions by @mattcaswell , but the latter activated only by presence of the DTLS1_3_VERSION define in the openssl code actually run. That way, the code should run and test OK under both conditions (feature/dtls1.3 landed or not).

**Comment by seschu** [comment]
> `oqsprovider` is meant to support "research and prototyping" (includes testing for me) and already contains a lot of code to that purpose -- primarily for PQC but why not for DTLS1.3?
> 
> So, I'm very open for that but leave this decision to @fwh-dc and/or @seschu : What's your preference/would it help you if we land DTLS1.3 (test) support in `oqsprovider` main?

Personally, I dont want to be the cause for too much "double" maintaining of branches :). If the DTLS1.3 support in oqsprovider would be added to main it would simplify my workflow a lot. However, I am not sure I understand what specifically you mean by "test" support.

**Comment by baentsch** [comment]
> it would simplify my workflow a lot

and mine :). See https://github.com/open-quantum-safe/oqs-provider/pull/586

> I am not sure I understand what specifically you mean by "test" support.

The (new) ability to run DTLS-specific (handshake) tests through all PQC algs (unlike only TLS1.3 tests up to now). Please check the PR: It may make it clearer what I mean. FYI, a manual run confirmed that all PQC KEM and SIG algs currently pass DTLS1.3 handshake tests after having activated the `BIO_s_dgram_mem` transport as suggested by @mattcaswell . So, as far as I can tell, this feature branch looks good also from a PQC perspective.

**Comment by fwh-dc** [comment]
Great news !

**Comment by c-southwest** [comment]
@fwh-dc Hello, I want to test your branch https://github.com/fwh-dc/oqs-openssl/tree/feature/dtls-1.3
But when I try to create an issue I found there is no "issue" tab.
Can you give me a suggestion where I can post issue? So I can ask you questions and provide wireshark captures.

**Comment by RAGHURH1** [comment]
when will DTLS 1.3 with pqc(kyber  or ml-kem) will be released (even if it is beta version)
can you provide me stimulated time (in months )

**Comment by fwh-dc** [comment]
Being the only contributer to the feature:

I do not know when it will be available. 

I haven't had ressources to work on it since march plus development was blocked until the acknowledge records pr was added to the feature branch.

Please feel free to browse for TODO(DTLS-1.3) on the feature branch for hints on missing functionality.

Summary:
All features for DTLS 1.3 has been implemented, but wee need to fix stability. I.e. what happens on dropped records fragments, make sure that TLS 1.3 features are available for DTLS 1.3 as well.

**Comment by baentsch** [comment]
@fwh-dc thanks for the status update. Sounds like a familiar issue (too many sensible things to do and too little time). One basic question, though: Given QUIC is now available in OpenSSL, what's the substantial advantage DTL1.3 would add (both are reasonably fast, datagram oriented TLS stacks, right)? Being neither QUIC nor DTLS expert, at a cursory look, QUIC to me actually seems to be designed for better performance, making the question more pronounced: What would be a good reason to bring the DTL1.3 branch to merge-readiness? Did the OpenSSL core team voice interest in this?

**Comment by Frosne** [comment]
> [@fwh-dc](https://github.com/fwh-dc) thanks for the status update. Sounds like a familiar issue (too many sensible things to do and too little time). One basic question, though: Given QUIC is now available in OpenSSL, what's the substantial advantage DTL1.3 would add (both are reasonably fast, datagram oriented TLS stacks, right)? Being neither QUIC nor DTLS expert, at a cursory look, QUIC to me actually seems to be designed for better performance, making the question more pronounced: What would be a good reason to bring the DTL1.3 branch to merge-readiness? Did the OpenSSL core team voice interest in this?

Isn't DTLS better suited for real-time media and such?

**Comment by fwh-dc** [comment]
> [@fwh-dc](https://github.com/fwh-dc) thanks for the status update. Sounds like a familiar issue (too many sensible things to do and too little time). One basic question, though: Given QUIC is now available in OpenSSL, what's the substantial advantage DTL1.3 would add (both are reasonably fast, datagram oriented TLS stacks, right)? Being neither QUIC nor DTLS expert, at a cursory look, QUIC to me actually seems to be designed for better performance, making the question more pronounced: What would be a good reason to bring the DTL1.3 branch to merge-readiness? Did the OpenSSL core team voice interest in this?

Our primary goal is to use it for DTLS-SRTP and SRTP for voice and video calls.

And other TLS implementations has DTLS 1.3 support.

**Comment by baentsch** [comment]
Thanks for the feedback @fwh-dc @Frosne . If not hearing objections, I'd like to try and use this issue to test out how the "new" OpenSSL TAC/BAC decision-making processes work by raising the question as to which wider interest exists for DTLS 1.3 by the community (e.g., usage & contributing new tests) and the committers (e.g., reviews) and how this is linked into GH interactions.

**Comment by fwh-dc** [comment]
Is the release 3.6 priorities poll relevant?
https://openssl-communities.org/p/8lj6EgwJ?stance_token=wgcsvYRa57BehPkHpLBon7dZ&utm_campaign=outcome_created&utm_medium=email

**Comment by baentsch** [comment]
Thanks for the pointer @fwh-dc ! I didn't know about it -- but maybe that's because I'm not a committer... I created https://openssl-communities.org/d/fskEj9eM/interest-in-dtls-1-3-support- in the hope of a wider (particularly user) community discussion -- but with some doubts as to whether this works as the "General Discussion" group seems to be even further "siloed away" from all other communities... Maybe time to open another discussion regarding (breaking down barriers between) those silos... :-)

**Comment by baentsch** [comment]
@fwh-dc If you didn't follow it already, there's quite some statements of support in the [OpenSSL communities discussion forum](https://openssl-communities.org/d/fskEj9eM/interest-in-dtls-1-3-support-). Please feel free to chime in there (and/or propose concrete next tasks here). My interest would clearly be enablement of PQC using OpenSSL's default providers -- given support via [oqsprovider exists since some time](https://github.com/open-quantum-safe/oqs-provider/pull/586) this seems like a modest task only, though.

**Comment by fwh-dc** [comment]
Following the meeting held on Tuesday last week during the OpenSSL Conference I am posting a DTLS 1.3 Feature Branch Status.

Completed Work:

* Implemented DTLS 1.3 specific features:
   * Record acknowledgements
   * Encrypted sequence numbering
   * etc.
* Some TLS 1.3 test suites duplicated and adapted for DTLS 1.3
* Some existing DTLS tests patched to support DTLS 1.3

Outstanding Work:

* Remaining TODO(DTLS-1.3) items in the codebase need to be addressed
* Some TLS 1.3 functionality is incompatible with DTLS 1.3 and requires patching, even under a limited initial scope (part of the TODO's)
* Further development needed for handling dropped fragments and messages
* Design documentation needs review — is the current coverage sufficient?
* Clarification needed on OpenSSL’s criteria for accepting DTLS 1.3 as mature for release

I currently have the following PR's:

* Support DTLS 1.3 in dtlstest.c: https://github.com/openssl/openssl/pull/26629
* Fix resumption in DTLS 1.3: https://github.com/openssl/openssl/pull/26796
* Include some tests that have been turned off: https://github.com/openssl/openssl/pull/26587
* Handle wrap around for the epoch correctly: https://github.com/openssl/openssl/pull/23553

I will also make additions to ossltest provider to allow the TLSProxy to record sequence numbers which are need for the DTLS ack test cases.

**Comment by rsith71** [comment]
I added a couple of tickets. The first two are to track some testing cleanup.

- SSL API Tests: https://github.com/openssl/project/issues/1667
- DTLS 1.3 MTU: https://github.com/openssl/project/issues/1668/

The other issue is I around Verifying the ACK of the New Session Ticket message sent. 
https://github.com/openssl/project/issues/1669

**Comment by fwh-dc** [comment]
FYI:
https://datatracker.ietf.org/doc/draft-ietf-tls-rfc9147bis/

I bis document has been created for DTLS 1.3. David Benjamin raised som concerns about ack'ing partial clienthello's (as far as I remember).

**Comment by mattcaswell** [comment]
PR to merge DTLSv1.3 support to master is in #32467. Once that is merged I think we can close this issue.

**Comment by mattcaswell** [comment]
DTLSv1.3 support is now in the master branch and will be in OpenSSL 4.1 to be released in October.

Thanks to everyone involved in this marathon effort, and @fwh-dc and @rsith71 in particular.

Closing this issue.
